### ISSUE TITLE
Tracking issue for the `thiscall` calling convention

### ISSUE THREAD

**Issue by froydnj** [rust-lang/rust#42202]
Added in #42058, which fixes #42044 and related requests from other bugs.

This calling convention is the default calling convention for C++ instance methods in the (MSVC) Windows ABI.  At the very least, bindgen (and its dependency syntex) need this to exist so they have a reasonable chance of representing C++ instance methods in an AST.  `thiscall` is therefore important for the interop story between Rust and C++ on Windows.

I don't *think* you could actually write callable C++ instance methods in Rust for MSVC, since mangled names in MSVC contain characters that, AFAICT, are not legal in Rust identifiers ('?', for instance).  But maybe there is cleverness around assigning assembly names to things that I am not aware of.

**Comment by ids1024** [comment]
> But maybe there is cleverness around assigning assembly names to things that I am not aware of.

If this doesn't exist, perhaps it warrants an RFC? In particular some kind of attribute to specify the symbol name as an alternative to `no_mangle`.

**Comment by froydnj** [comment]
> If this doesn't exist, perhaps it warrants an RFC? In particular some kind of attribute to specify the symbol name as an alternative to no_mangle.

I think this sounds like an excellent idea!

**Comment by ids1024** [comment]
It looks like there's already an `export_name` attribute that does this. That probably would work.

**Comment by steveklabnik** [comment]
Triage: not aware of any movement on stabilizing this.

**Comment by whatisaphone** [comment]
>not aware of any movement on stabilizing _this_

I see what you did there, and I like it.

More seriously, I'm just swinging by to say I use this feature independently of bindgen, and I hope it's stabilized one day (or at least sticks around in some form)

**Comment by MSxDOS** [comment]
What's blocking stabilization of this?

**Comment by pravic** [comment]
The same question.

cc @dtolnay to route this to an appropriate person.

**Comment by super-continent** [comment]
Wondering the same myself, any updates on stabilizing this? Would be very nice to have access to in stable.

**Comment by zorftw** [comment]
It's been rather long, why is still unstable?

**Comment by chorman0773** [comment]
Indeed. I have a use for this in my compiler project https://github.com/LightningCreations/lccc, which would allow me to write bindings to the C++ api in rust on MSVC (which I am able to do on other major compilers, but need thiscall for i386 windows).

**Comment by DrMeepster** [comment]
I think this feature should be stabilized.
# Stabilization Report
## Summary
The `abi_thiscall` feature allows code to export and link to functions with the thiscall abi. This is the abi used by default for C++ member functions on x86 MSVC, and is very important for C++ FFI.
## Tests
This feature is tested in [`ui/extern/extern-thiscall.rs`](https://github.com/rust-lang/rust/blob/master/src/test/ui/extern/extern-thiscall.rs)
## Documentation PRs
* The Reference: rust-lang/reference#1092
* The Nomicon: rust-lang/nomicon#311
There are no known issues or unresolved questions related to this feature.

**Comment by x87** [comment]
I have a library that builds for two targets: `i686-pc-windows-msvc` and `x86_64-pc-windows-msvc`. In this library there are two functions defined as:

```
pub extern "thiscall" fn foo(a: usize, b: usize) {}
pub extern "fastcall" fn bar(a: usize, b: usize) {}
```

they were compiling and working perfectly fine on Rust 1.57.

Now I get the hard error on `foo` when compiling to x86_64, saying
```
`"thiscall"` is not a supported ABI for the current target (rustc E0570)
```

What I find inconsistent here is that both functions have x86-only calling conventions (`thiscall` and `fastcall`), whereas the compiler only complains to the former. Also an error on the calling convention in x64 mode seems to be irrelevant as the compiler always uses (or should use) the x64 calling convention regardless of what was put in the declaration.

I can't find a way to satisfy the compiler without duplicating the function:
```
#[cfg(target_arch = "x86")]
pub extern "thiscall" fn foo(a: usize, b: usize) {}

#[cfg(target_arch = "x86_64")]
pub extern "fastcall" fn foo(a: usize, b: usize) {}
```

which adds unnecessary complexity to the code that was working just yesterday.

Is there a simpler way to support different calling conventions on the same function? If not, is there a strong argument to keep this as a hard error as opposed to a warning similar to other cases (fastcall et al)? Thanks

**Comment by DrMeepster** [comment]
@x87 You can write a [tt muncher macro](https://danielkeep.github.io/tlborm/book/pat-incremental-tt-munchers.html) to set the abi, but I don't think that should be needed. A better solution would be to allow macros in abi position, but I'm unsure if an RFC is needed for this.

**Comment by pnkfelix** [comment]
Discussed at T-compiler backlog bonanza.

We're thrilled to see such interest and even a full stabilization report.

We are concerned about the issues raised about the current error behavior from using the "thiscall" ABI on non i686 targets.

We are specifically interested in the answer to the question raised by @x87: Why is "thiscall" a hard error while other cases (fastcall et al) are just warnings?

And we might also be interested in working out an ergonomic way to write cfg-dependent ABIs, which probably does require allowing macros in ABI position, as mentioned by @DrMeepster 

@rustbot label: S-tracking-design-concerns S-tracking-needs-summary

**Comment by petrochenkov** [comment]
>Why is "thiscall" a hard error while other cases (fastcall et al) are just warnings?

Because it's unstable, and other cases are stable and could not be made hard errors due to backward compatibility.

**Comment by petrochenkov** [comment]
There are two ways to see `thiscall`:
- A concrete ABI that only makes sense on i686 and is used for C++-style method calls (similar to `extern "stdcall"`).
- A generic ABI for C++-style method calls that is equivalent to `extern "C"` everywhere except for i868 (similar to `extern "system"`). This is a pure backward-compatible extension to the first alternative.

Right now the first alternative is implemented, and can be stabilized as is, because the second alternative is a pure extension.

The second alternative, however, is much more useful for tools like bindgen that need to generate Rust versions of C++-style methods in a cross-platform way (same generated code for all targets).
So it can also be implemented and stabilized, either separately from the first alternative or simultaneously, it doesn't matter much.

**Comment by nagisa** [comment]
> Also an error on the calling convention in x64 mode seems to be irrelevant as the compiler always uses (or should use) the x64 calling convention regardless of what was put in the declaration.

This is not true for `vectorcall`. I also believe Microsoft engineers made a mistake by implementing this behaviour in the first place, and I think it was mistake for us to adopt this same behaviour in https://github.com/rust-lang/rust/pull/54576. I would rather we don't make the same mistake for `thiscall`, given that our hand is not being yet forced by the stability promise here.

**Comment by ChrisDenton** [comment]
In practice, is there any time when developers would not want this fallback behaviour? IIRC, pretty much the reason for using these ABIs is for interop with C/C++ APIs.

**Comment by nagisa** [comment]
> pretty much the reason for using these ABIs is for interop with C/C++ APIs

I have used specific calling conventions (though not `thiscall` in particular) to call into hand written assembly code because they made dealing with calling conventions easier. The fact that these can become something else than what's specified in the code was and remains a major footgun in that codebase (which is somewhat mitigated by the symbol mangling some of these calling conventions imply, but still…)

**Comment by ChrisDenton** [comment]
Right but I mean specifically "thiscall" and "stdcall" on Windows platforms. On other platforms I'm not at all against using them being a hard error. To be honest, my worry at the moment is the potential for making the common case more complicated and error prone in order to make it simpler for a theoretical case that nobody actually needs.

Lastly I'd point out that "C" is even more magic because it changes depending on way more than if it's 32-bit x86 or not.

**Comment by nagisa** [comment]
> Right but I mean specifically "thiscall" and "stdcall" on Windows platforms. [...] Lastly I'd point out that "C" is even more magic because it changes depending on way more than if it's 32-bit x86 or not.

I used `fastcall` in [`psm`](https://github.com/rust-lang/stacker/blob/547807f03246783ada94b956a575b9ce257dc2ca/psm/src/lib.rs#L36-L44), which is neither `thiscall` nor `stdcall`, but is largely in the same camp. And, indeed, a primary motivation for its use was to not have to deal with the C calling convention differences across platforms.

> To be honest, my worry at the moment is the potential for making the common case more complicated and error prone in order to make it simpler for a theoretical case that nobody actually needs.

The way I see it the very fact that specifying `fastcall/stdcall/vectorcall/etc` has a potential to not give a programmer exactly a `fastcall/stdcall/vectorcall/etc` is wrong from the first principles in the context of Rust.

There are plenty of other scenarios where Rust could make it more straightforward to write code, following precedent set by others. Automatic operand conversion (in JavaScript expressions like `"5" + "3"` produce `"53"`, but `"5" - "3" => 2`) or boolean operand coercion are the relevant examples that come to my mind most readily. Rust was largely designed around absence of this sort of magic precisely because such magic being error-prone was ultimately one of Rust’s takeaways from the giants that came before it.

---

Side note: I also want to draw the attention to how careful Microsoft is in their wording to avoid specifying this behaviour as an inherent property of the calling convention itself. Similar verbiage can be observed in other MSDN pages.

> In most cases, keywords or compiler switches that specify an unsupported
> convention on a particular platform are ignored, and the platform default
> convention is used. – [MSDN](https://docs.microsoft.com/en-us/cpp/cpp/argument-passing-and-naming-conventions)

> On ARM, ARM64, and x64 machines, __thiscall is accepted and ignored by the compiler. – [MSDN](https://docs.microsoft.com/en-us/cpp/cpp/thiscall?view=msvc-160)

**Comment by workingjubilee** [comment]
x64 is a particularly annoying/amusing case, because while the default is a "four-register fast-call calling convention" (I imagine MSDN **winking** here), it is not actually a perfect match with `__fastcall` and I would feel dubious about calling it a superset of `__fastcall`.

**Comment by codecheck01** [comment]
Any updates on getting this to stable Rust?

**Comment by oberien** [comment]
Today marks the 6 year anniversary of this issue. In 2 days is the 6 year anniversary [of merging `thiscall` into (nightly) rust](https://github.com/rust-lang/rust/pull/42058#event-1098702482).
* There has been a [Stabilization Report](https://github.com/rust-lang/rust/issues/42202#issuecomment-922652127).
* It's been [discussed at T-compiler backlog bonanza](https://github.com/rust-lang/rust/issues/42202#issuecomment-1072485606) with one question raised.
* The question has been [responded to and answered](https://github.com/rust-lang/rust/issues/42202#issuecomment-1072538404).
* There are 90 projects on GitHub using `feature(abi_thiscall)` according to [a quick github search](https://github.com/search?q=feature%28abi_thiscall%29+language%3ARust&type=code&l=Rust)
* I've personally been using this feature successfully [since Nov 2017](https://github.com/oberien/refunct-tas/commit/10d7597977de56f2e79bd83d76522c55d4f3e8cb#diff-e15a86cee792774f4682f3afa1238285334185ce8e3ab3e86120535a408e771eR5) for compatibility with Unreal Engine on Windows (whose default Windows compilation target is 32bit using `thiscall` for all class methods).

Thus I'd like to ask: Is there anything blocking the stabilization of the `abi_thiscall` feature? If yes, how can we help move its stabilization forward?

**Comment by DrMeepster** [comment]
@oberien I believe that [this](https://github.com/rust-lang/rust/issues/42202#issuecomment-1006213947) is the main blocker. Allowing macro expansions in ABI strings was proposed to fix this (rust-lang/lang-team#151), but that was rejected.

**Comment by petrochenkov** [comment]
>I believe that [this](https://github.com/rust-lang/rust/issues/42202#issuecomment-1006213947) is the main blocker. 

That's not a blocker, the stabilization is not currently blocked on anything, just on someone submitting a PR and stabilization report.
The fix for https://github.com/rust-lang/rust/issues/42202#issuecomment-1006213947 is a backward compatible extension.

**Comment by DrMeepster** [comment]
Alright then, here's a slightly updated stabilization report. I think this can be FCP'd?

# Stabilization Report
## Summary
The `abi_thiscall` feature allows code to export and link to functions with the thiscall abi. This is the abi used by default for C++ member functions on x86 MSVC, and is very important for C++ FFI.
## Tests
This feature is tested in [`ui/extern/extern-thiscall.rs`](https://github.com/rust-lang/rust/blob/master/tests/ui/extern/extern-thiscall.rs)
## Documentation PRs
* The Reference: rust-lang/reference#1092
* The Nomicon: rust-lang/nomicon#311
## Known issues
* `thiscall` is somewhat unergonomic for code that works on both x86 and x64 (https://github.com/rust-lang/rust/issues/42202#issuecomment-1006213947)
  *  This doesn't block stabilization because a fix would be backward compatible

**Comment by x87** [comment]
FWIW, I solved [my issue](https://github.com/rust-lang/rust/issues/42202#issuecomment-1006213947) using [this crate](https://crates.io/crates/fn_abi). It does exactly what I would expect from the compiler itself. 

So I may call it unergonomic, but not a blocker, for sure.

**Comment by Trolldemorted** [comment]
How does one trigger an FCP? After reading through the [stablization guide](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#fcp), "any member of the team responsible for tracking this feature" has to start the process by commenting `@rfcbot fcp merge`. According to the tags of this issue and the compiler team member list [here](https://www.rust-lang.org/governance/teams/compiler), @petrochenkov would be eligible for that.

@petrochenkov could you either do that, or tell us what needs to be done? Right now we have to call thiscall functions with `asm!`, which is very annoying and dangerous since one might get the clobbering wrong.

**Comment by petrochenkov** [comment]
@rfcbot fcp merge
Stabilization report: https://github.com/rust-lang/rust/issues/42202#issuecomment-1565207948

**Comment by jackh726** [comment]
Not sure if this should, at the very least, *also* include T-lang in the FCP. Quickly looking at previous ABI stabilizations, it's been tagged either T-compiler/T-lang, or T-lang alone, e.g. #36167 and #65815.

**Comment by soloist-v** [comment]
> 

The use of thiscall is necessary in x64 mode because bindgen would generate the incorrect binding for the C++ member function Dims getBindingDimensions(int32_t bindingIndex) const;. It would generate:
```rust
extern "C" {
    #[link_name = "\u{1}?getBindingDimensions@ICudaEngine@nvinfer1@@QEBA?AVDims32@2@H@Z"]
    pub fn ICudaEngine_getBindingDimensions1(
        this: *const root::nvinfer1::ICudaEngine,
        bindingIndex: i32,
    ) -> root::nvinfer1::Dims;
}
```
Clearly, this result is incorrect as the this pointer will be passed through rdx (The this pointer should be passed through rcx). The correct result should be:
```rust
extern "C" {
    #[link_name = "\u{1}?getBindingDimensions@ICudaEngine@nvinfer1@@QEBA?AVDims32@2@H@Z"]
    pub fn ICudaEngine_getBindingDimensions1(
        this: *const root::nvinfer1::ICudaEngine,
        dims: *mut root::nvinfer1::Dims,
        bindingIndex: i32,
    );
}
```
A better approach (in x64 mode) would be:
```rust
extern "thiscall" {
    #[link_name = "\u{1}?getBindingDimensions@ICudaEngine@nvinfer1@@QEBA?AVDims32@2@H@Z"]
    pub fn ICudaEngine_getBindingDimensions1(
        this: *const root::nvinfer1::ICudaEngine,
        bindingIndex: i32,
    ) -> root::nvinfer1::Dims;
}
```

**Comment by nagisa** [comment]
Yeah, this definitely sounds like a T-lang question than a T-compiler one. What would be in scope of the T-compiler here is an advice to T-lang on the implementation quality of this ABI (which is no better or worse than the other stable ABI options people can pick from today.)

**Comment by Trolldemorted** [comment]
Should we @ them directly, or will they notice because it is now tagged T-lang?

**Comment by compiler-errors** [comment]
I don't think T-lang can cancel an FCP that they're not participating. I can nominate it and cc @rust-lang/lang. Happy to cancel the FCP in favor if they feel like it's necessary.

**Comment by joshtriplett** [comment]
@rfcbot poll T-lang Should we stabilize thiscall?

**Comment by nikomatsakis** [comment]
Discussed in meeting. We are generally in favor of this. To avoid procedural pain we decided to just use rfcbot poll.

**Comment by compiler-errors** [comment]
canceling fcp in favor of poll above

@rfcbot fcp cancel

**Comment by compiler-errors** [comment]
@rfcbot cancel

**Comment by joshtriplett** [comment]
(The T-lang poll was meant to substitute for T-lang's inclusion in a multi-team FCP. I hope this wasn't taken as a *substitute* for making sure T-compiler was fine with this too.)

**Comment by compiler-errors** [comment]
I interpreted:

1. the reassignment of team tag to T-lang by @nagisa
2. @oli-obk's r+ in #114562

as implicitly saying "T-lang's decision, not ours".

At this point, if any @rust-lang/compiler team member thinks that we should have continued with that FCP, they should speak up, since we'll have to do a retroactive FCP at this point (and/or unland that PR). I *personally* think that an FCP is unnecessary from our side, though.
